conans/client/toolchain/meson.py

```diff
@@ -0,0 +1,176 @@
+        self._fpic = self._conanfile.options.get_safe("fPIC")
+        self.definitions = dict()
+
+    @staticmethod
+    def _to_meson_value(value):
+        # https://mesonbuild.com/Machine-files.html#data-types
+        import six
+
+        try:
+            from collections.abc import Iterable
+        except ImportError:
+            from collections import Iterable
+
+        if isinstance(value, six.string_types):
+            return "'%s'" % value
+        elif isinstance(value, bool):
+            return 'true' if value else "false"
+        elif isinstance(value, six.integer_types):
+            return value
+        elif isinstance(value, Iterable):
+            return '[%s]' % ', '.join([str(MesonToolchain._to_meson_value(v)) for v in value])
+        return value
+
+    @staticmethod
+    def _to_meson_build_type(build_type):
+        return {"Debug": "'debug'",
+                "Release": "'release'",
+                "MinSizeRel": "'minsize'",
+                "RelWithDebInfo": "'debugoptimized'"}.get(build_type, "'%s'" % build_type)
+    # FIXME : use 'custom' otherwise? or use just None?
+
+    @property
+    def _debug(self):
+        return self._build_type == "Debug"
+
+    @property
+    def _ndebug(self):
+        return self._build_type != "Debug"
+
+    @staticmethod
+    def _to_meson_vscrt(vscrt):
+        return {"MD": "'md'",
+                "MDd": "'mdd'",
+                "MT": "'mt'",
+                "MTd": "'mtd'"}.get(vscrt, "'none'")
+
+    @staticmethod
+    def _to_meson_shared(shared):
+        return "'shared'" if shared else "'static'"
+
+    def _to_meson_cppstd(self, cppstd):
+        if self._base_compiler == "Visual Studio":
+            return {'14': "'vc++14'",
+                    '17': "'vc++17'",
+                    '20': "'vc++latest'"}.get(cppstd, "'none'")
+        return {'98': "'c++03'", 'gnu98': "'gnu++03'",
+                '11': "'c++11'", 'gnu11': "'gnu++11'",
+                '14': "'c++14'", 'gnu14': "'gnu++14'",
+                '17': "'c++17'", 'gnu17': "'gnu++17'",
+                '20': "'c++1z'", 'gnu20': "'gnu++1z'"}.get(cppstd, "'none'")
+
+    @staticmethod
+    def _none_if_empty(value):
+        return "'%s'" % value if value.strip() else None
+
+    @property
+    def _native_content(self):
+        project_options = []
+        for k, v in self.definitions.items():
+            project_options.append("%s = %s" % (k, self._to_meson_value(v)))
+        project_options = "\n".join(project_options)
+
+        context = {
+            # https://mesonbuild.com/Machine-files.html#project-specific-options
+            "project_options": project_options,
+            # https://mesonbuild.com/Builtin-options.html#directories
+            # TODO : we don't manage paths like libdir here (yet?)
+            # https://mesonbuild.com/Machine-files.html#binaries
+            "c": self._none_if_empty(os.environ.get("CC", '')),
```

Comment:
I would try to have the ``os.environ`` abstracted away. Pass an "env" to the toolchain constructor or something like that. If not provided, it will use ``os.environ``, but at this point you already get an abstraction.

Reply:
added env to the ctor